### PR TITLE
[Feature] Tvm integration

### PR DESCRIPTION
**Pull request by kira-lin**
## Description
This branch enables dgl to call TVM AOT compiled modules. It adds a Tree reduction SDDMM as an example.

## Checklist
- [x] SDDMM support more data types
- [x] handle loading libfeatgraph.so properly

### CODE REVIEW

File: cmake/tvm/pack_featgraph.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import os
+from sddmm_tr import sddmm_tree_reduction
+
+lib_path = '../../build/libfeatgraph.so'
```
Comment (yzh119) [inline cmake/tvm/pack_featgraph.py:4]: The path is hardcoded, we should create a function `find_lib_path` like https://github.com/dmlc/dgl/blob/ad8530a4dc2722e481f56c1d417c36cd87bbfb19/python/dgl/_ffi/libinfo.py#L8 to locate the libpath.

**Review by yzh119 — COMMENTED** [review]
LGTM

**Comment by yzh119** [comment]
Looks like there is a compilation error, @kira-lin would you mind double checking it?

**Comment by yzh119** [comment]
@jermainewang The PR is now ready for review.

File: featgraph/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+include(../cmake/util/FindCUDA.cmake)
+find_cuda(ON)
+message(STATUS "${CUDA_INCLUDE_DIRS}")
+
+add_custom_target(
+  featgraph_kernel
+  COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/pack_featgraph.py
+)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -O2 -fPIC")
```
Comment (jermainewang) [inline featgraph/CMakeLists.txt:16]: Is 14 required?

File: featgraph/include/featgraph.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#include <dlpack/dlpack.h>
+
+namespace dgl {
+namespace featgraph {
+
+void SetFeatGraphModulePath(const std::string& path);
+
+void SDDMMTreeReduction(DLManagedTensor* row, DLManagedTensor* col, 
+                        DLManagedTensor* lhs, DLManagedTensor* rhs, 
+                        DLManagedTensor* out);
```
Comment (jermainewang) [inline featgraph/include/featgraph.h:18]: pls add some docstring

File: featgraph/pack_featgraph.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    List[IRModule]:
+        The list of IRModules.
+    """
+    ret = []
+    # SDDMM Tree Reduction
+    for dtype in dtypes:
+        for idtype in idtypes:
+            ret.append(sddmm_tree_reduction_gpu(idtype, dtype))
+
+    # TODO(zihao): SDDMM Naive Solution
```
Comment (jermainewang) [inline featgraph/pack_featgraph.py:26]: what's this todo?

File: featgraph/src/featgraph.cc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+/* \brief Singleton that loads the featgraph module. */
+class FeatGraphModule {
+public:
+  static FeatGraphModule* Global() {
+    static FeatGraphModule inst;
+    return &inst;
+  }
+
+  void SetPath(const std::string& path) {
```
Comment (jermainewang) [inline featgraph/src/featgraph.cc:23]: `SetPath` -> `Load`

File: featgraph/src/featgraph.cc [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    return ret;
+  }
+private:
+  tvm::runtime::Module mod;
+  FeatGraphModule() {}
+};
+
+/* Set the path of libfeatgraph_kernels.so */
+void SetFeatGraphModulePath(const std::string& path) {
```
Comment (jermainewang) [inline featgraph/src/featgraph.cc:41]: `LoadFeatGraphModule`

**Review by jermainewang — APPROVED** [review]
LGTM. some minor comments. Great job! @kira-lin and @yzh119

File: featgraph/CMakeLists.txt (hunk 2 above)
Comment (yzh119) [reply featgraph/CMakeLists.txt:17]: TVM runtime requires C++14 to build.

**Pull request merged**